### PR TITLE
Cursor unit tests

### PR DESCRIPTION
**Pull request by jenn-le**
- Adds tests for general json-compatible cursor implementations based on the existing JsonCursor tests
- Ports tests to new cursor API
- Adds additional tests specific to the new API

### CODE REVIEW

**Comment by CraigMacomber** [comment]
Did you check that the enabled assert's performance impact?

File: packages/dds/tree/src/test/cursorLegacy.spec.ts [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+import { strict as assert } from "assert";
+
+import { ITreeCursor, TreeNavigationResult } from "../forest";
+import { EmptyKey, FieldKey, JsonableTree } from "../tree";
+import { brand } from "../util";
+import { cursorTestCases } from "./cursor.spec";
+
+/**
+ * Tests the provided cursor factory with JsonableTree data. The cursor must be JSON compatible.
+ * @param suiteName - The name of the test suite to create.
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/cursorLegacy.spec.ts:15]: this is uses as part of the suiteName. Maybe "cursorName" or just "name" would be better?

File: packages/dds/tree/src/tree/cursor.ts [hunk 2]
```diff
@@ -218,6 +218,33 @@ export interface ITreeCursorSynchronous extends ITreeCursor {
...
 
+/**
+ * @param cursor - tree whose fields will be visited.
+ * @param f - builds output from field, which will be selected in cursor when cursor is provided.
+ * If `f` moves cursor, it must put it back to where it was at the beginning of `f` before returning.
+ * @returns array resulting from applying `f` to each field of the current node on `cursor`.
+ * Returns an empty array if the node is empty or not present (which are considered the same).
+ * Note that order is not specified for field iteration.
+ */
+ export function mapCursorFields<T>(cursor: ITreeCursor, f: (cursor: ITreeCursor) => T): T[] {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/tree/cursor.ts:229]: these kinds of functions should take in a generic cursor type, and provide it to the callback so they can be used with Forest specific cursors are needed.

File: packages/dds/tree/src/test/cursor.spec.ts [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+                    getKeysAsSet(factory({
+                        type: brand("Foo"),
+                        fields: { [EmptyKey]: [{ type: brand("Bar"), value: "test" }, { type: brand("Bar") }] },
+                    })),
+                    new Set([EmptyKey]),
+                );
+            });
+
+            it("string", () => {
+                assertSetsEqual(getKeysAsSet(factory({ type: brand("Foo"), value: "" })), new Set([]));
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/cursor.spec.ts:144]: I believe assert.deepEqual handles sets, and will produce much nicer assert messages showing diffs than this helper.

Also, `new Set()` does not need and empty array.

**Comment by jenn-le** [comment]
> Did you check that the enabled assert's performance impact?

Good call. I checked and there's a minimal difference where some runs are actually faster with the asserts so I don't think it matters too much.

**Review by CraigMacomber — APPROVED** [review]

**Pull request merged**